### Running a forward pass

What is a forward pass?
- When input data flows through a neural network in a forward direction to produce outputs or prediction, it passes or flows through each network layers
- Calculations transform the data into new representation at every layer, which are passed to the next layer, until the final layer generates **outputs**

> The purpose of the ***forward pass*** is to pass input data through the network and produce predictions or outputs based on the model's learned parameters, also known as **weights** and **biases**


> This process is essential for both training and making new predictions

The final output can be:
- Binary classification
- Multi-class classification
- Numerical Predictions or Regressions

---

**Binary Classification: Forward Pass**

In [2]:
# Create input data of shape: 5x6
import torch
import torch.nn as nn

input_data = torch.tensor(
    [[-0.4421, 0.1234, 0.5678, -0.9876, 0.5432, -0.1111],
     [0.2345, -0.6789, 0.1111, 0.0322, -0.3333, 0.4444],
     [-0.5555, 0.6866, -0.8777, 0.8848, -0.7899, 1.0000],
     [1.1671, -1.2222, 1.3333, -1.4444, 1.5555, -1.6666],
     [1.7577, -1.8888, 1.9999, -2.0000, 2.1111, -2.2253]]
)

# 5 animal x 6 feature data

In [3]:
# Create binary classification model (through the use of a sigmoid activation function)
model = nn.Sequential(
    nn.Linear(6, 4),  # Input layer: 6 features, Output layer: 4 neurons
    nn.Linear(4, 1),  # Hidden layer: 4 neurons, Output layer: 1 neuron
    nn.Sigmoid()      # Activation function for output layer
)

In [4]:
# Pass input data through model
output = model(input_data)
print(output)

tensor([[0.5280],
        [0.4945],
        [0.6173],
        [0.4090],
        [0.3411]], grad_fn=<SigmoidBackward0>)


Since this is a ***sigmoid activation function***, the output is binary:
- Five (5) probabilities between 0 and 1, one for each animal
- Classification (0.5 Threshold):
    - Class = 1 (mammal) for values >= 0.5 (0.5280, 0.6173)
    - Class = 0 (not mammal) for values < 0.5 (0.3411, 0.4090, 0.4945)

---

**Multi-class Classification: Forward Pass**
> Class 1 - Mammal, Class 2 - Bird, Class 3 - Reptile

In [6]:
n_classes = 3

# Create a multi-class classification model
model = nn.Sequential(
    nn.Linear(6, 4),
    nn.Linear(4, n_classes), 
    nn.Softmax(dim=-1)
)

# Pass input data through model
output = model(input_data)
print(output)

tensor([[0.2647, 0.4838, 0.2515],
        [0.3215, 0.2792, 0.3992],
        [0.3657, 0.2447, 0.3896],
        [0.1864, 0.5544, 0.2592],
        [0.1509, 0.6191, 0.2299]], grad_fn=<SoftmaxBackward0>)


Since this is a multi-class classification model, each *row* sums to ***one***.
- Predicted label = class with the highest probability
- Row 1 (0.4838) = class 2 (not mammal)
- Row 2 ( 0.3992) = class 3 (reptile)
- Row 3 (0.3896) = class 3 (reptile)
- Row 4 (0.5544) = class 2 (not mammal)
- Row 5 (0.6191) = class 2 (not mammal)

---

**Regression: Forward Pass**
> Predicting continuous numerical value

In [7]:
# Create regression model

model = nn.Sequential(
    nn.Linear(6, 4),
    nn.Linear(4, 1)
)

# pass input data through model
output = model(input_data)
print(output)

tensor([[-0.2340],
        [ 0.2416],
        [ 0.2650],
        [-0.4180],
        [-0.5525]], grad_fn=<AddmmBackward0>)
